packages/mui-material/src/TableSortLabel/TableSortLabel.js

```diff
@@ -126,11 +140,7 @@ const TableSortLabel = React.forwardRef(function TableSortLabel(inProps, ref) {
     >
       {children}
       {hideSortIcon && !active ? null : (
-        <TableSortLabelIcon
-          as={IconComponent}
-          className={clsx(classes.icon)}
-          ownerState={ownerState}
-        />
+        <TableSortLabelIcon as={IconSlot} className={clsx(classes.icon)} {...iconProps} />
```

Comment:
Shouldn't this be:

```jsx
  const externalForwardedProps = {
    slots: {
        icon: IconComponent,
        ...slots
    },
    slotProps,
  };

  const [IconSlot, iconProps] = useSlot('icon', {
    elementType: TableSortLabelIcon,
    externalForwardedProps,
    ownerState,
    className: classes.icon
  });

// ...
  <IconSlot {...iconProps} />
```

Reply:
I just realized we did the same in https://github.com/mui/material-ui/pull/41875/files#diff-d626958b0fab135d0026b2cf86b08f733dae8abb98fc406df247b3846dcf8589R651

We will need to fix that 🤔

Reply:
are you sure?

By going with your approach `TableSortLabelIcon` won't be in consideration by this logic. 

https://github.com/mui/material-ui/blob/93e25916268e33832e78a10e3b822c25f6902a56/packages/mui-material/src/utils/useSlot.ts#L117

slots[name] (here: IconComponent) is preceeds over initialElementType ( here: TableSortLabelIcon)

Reply:
Oh, I see what's happening. There are some `*Component` props that are used as

```jsx
<DefaultComponent
    as={ComponentProp}
/>
```

But others are used as

```jsx
const { ComponentProp: DefaultComponent } = props

// ...

<ComponentProp />
```

Which is different.

If I'm not mistaken, the convention for the slot pattern is to work as the latter: When you provide a slot, it replaces the default completely. Is this correct @siriwatknp?

If this is correct, then we would have to:

- Deprecate `IconComponent` and keep providing it through `as`. 
- Implement `slots.icon` so that it replaces the slot completely (this is different from what we're doing now), following the pattern
- In the deprecation message we would have to acknowledge that `slots.icon` does not work exactly as `IconComponent`

Reply:
Other components with the `as` behavior:

- [PaginationItem](https://github.com/mui/material-ui/pull/41777/files#diff-014e80cab0c8e6df4ca350c7fd91d2d7b490250af715108435ed9113e1886216R412)
- [Autocomplete](https://github.com/mui/material-ui/pull/41875/files#diff-d626958b0fab135d0026b2cf86b08f733dae8abb98fc406df247b3846dcf8589R651)

Reply:
Discussed before in here: https://github.com/mui/material-ui/issues/34334#issuecomment-1251936061

Reply:
Let's wait for @siriwatknp's response, but my proposal would be:

- Deprecate `IconComponent` and keep providing it through `as`, so we keep the same functionality. 
- Implement `slots.icon` so that it replaces the slot completely, following the pattern
- In the deprecation message, the correct replacement for `IconComponent` would be `slotProps.icon.as`

Reply:
> In the deprecation message, the correct replacement for IconComponent would be slotProps.icon.as

I assume codemod should also do same

Reply:
> I assume codemod should also do same?

Yes, but let's wait for confirmation

Reply:
For this case, it should not be the `IconSlot` but `SortLabelIcon` slot?

```js
const [SortLabelIcon, sortLabelIconProps] = useSlot('sortLabelIcon', { elementType: TableSortLabelIcon, externalForwardedProps, ownerState, className: classes.icon, … });
```

Reply:
> I don't think the last is needed. Currently, we don't document as prop for Material UI, only component prop.

But what should we say the replacement for `IconComponent` is? Options:

1. We could say the replacement is `slots.icon` but it's not the same (replaces entire slot)
2. We could implement the `icon` prop which behaves as `IconComponent` does right now. I think it's a common-enough case to change the icon to justify a dedicated prop.

@siriwatknp

Reply:
Hey @DiegoAndai @siriwatknp any discussion happened on this PR front ?

Reply:
Sorry for the delay @sai6855, see: https://github.com/mui/material-ui/issues/41281#issuecomment-2542142649

For this case, it means we shouldn't deprecate `IconComponent`. It's not one of the props that should be moved to the slots pattern.

Reply:
Sorry for the delay @sai6855. I think this is a special case (the same for other `IconComponent` props in the codebase). Let's not deprecate.